### ISSUE TITLE
Mapillary Blurs

### ISSUE THREAD

**Issue by AnkEric** [mapillary/mapillary_issues#2846]
Mapillary Blurs have been discussed before, several times.
Sorry for repeating...

I am wondering why - so many - incorrect blurs are being set:

- branches (in trees)
- leaves (in trees)
- clouds (in the sky)
- birds (in the sky)
- window frames (in houses)
- gates, fences
- paving_stones
- scrub (nature)
- wheels (of a motorcar)
- Information boards having text only, so 80% of image is text (dozens of blurs are no exception)
- traffic_signs

![image](https://user-images.githubusercontent.com/20150550/34676716-112db04a-f48e-11e7-8c3c-000c8a12797f.png)

Does our unblurring make the algorithm smarter?

I am wondering why - so many - blurs are missed:

- license plates
- license plates not having the expected plate size (square)
- faces and faces and faces... 
- faces on short distance (so these are large...)

Examples...

![image](https://user-images.githubusercontent.com/20150550/34675485-12713462-f48a-11e7-84ba-72f7af885bd9.png)

Dark image is no argument.
Yesterday I encountered this image. Blurs have been set by me yesterday. Manually.

![image](https://user-images.githubusercontent.com/20150550/34676118-0dfd053a-f48c-11e7-8e54-e934a3e4b791.png)

But... I also missed a Blur justification. Mapillary spotted a face inside a restaurant behind a window pane!
Excellent! (Not for me, for Mapillary).

Also blurring procedure has room for improvement.

Suggested Buttons:
- Next image (without any changes)
- Next image having blurs set (filter on blurs)
- keyboard shortcuts
For instance:
Spacebar = Next
Enter = "Clear All +>" (no changes) or "Submit +>" (changes have been made by user).

![image](https://user-images.githubusercontent.com/20150550/34796450-d8757246-f655-11e7-9527-32736d341f28.png)

Also: I do make mistakes, often: "Clear All +>".
In the exact split second I hit the button I see a correct blur on a license plate. Which I just unblurred by mistake.
Stupid?! Oké, but very hard to correct. To be honest: I usually do not correct.

**Comment by Alessandro-Barbieri** [comment]
I noticed that the algorithm blurs only perfectly readable and perfectly horizontal license plates.

**Comment by GITNE** [comment]
@AnkEric Although these have been answered before or elsewhere, and some of these questions you could be answer yourself if would take the time and effort to think about them, I am going to explain them briefly.
> * Next image (without any changes)

Already exists. These are the `<` and `>` buttons. Alternatively, you can use `Alt` + `↑` and `Alt` + `↓`.

> Next image having blurs set (filter on blurs)

Seems rational at first but the intention and purpose of the blur editor is to not only correct false positives but also false negatives. A `Jump to next image with blurs` button would elude most users to only fix false positives, and perhaps only few false negatives which initially had some false positives in the same image.

> Spacebar = Next

Bad idea. The `SPACE` key is already a reserved global hot key in most browsers for page down.

> Enter = "Clear All +>" (no changes) or "Submit +>" (changes have been made by user).

Also bad idea. The `ENTER/RETURN` key heavily depends on browser or page component focus and has also the potential to collide with default inputs (or answers to) to browser, system, and other application pop-ups.

In general, better recognition demands more training which AFAICT Mapillary is about to schedule with the next major backend update. But, I think @peterneubauer can shed more light on this. :wink:

**Comment by AnkEric** [comment]
> Next image (without any changes)
> Already exists. These are the < and > buttons.

I know, see my image showing 3 arrows pointing to 3 controls.

> Alternatively, you can use Alt + ↑ and Alt + ↓.

That would be a solution. However [Alt ←], [Alt →] does not act as expected in Google Chrome.

For Blurs we need four controls (already existing):
- Back
- Forward
- Submit + Forward
- Clear All + Forward

Two options:
- Using mouse I want to move over short (!) distance to either of them and all four need to be large in size (and equal in  size).
Suggested: 4 large controls in the right side taskbar (identical to the 2 already located here).
Now hitting [forward] means: mouse move over 30 centimeters and concentrating to hit the right - small - spot. I’m not a gamer...
- Using keyboard I want to use keyboard shortcuts. You are correct: spacebar and enter is bad suggestion. Oké suggest other combination like: [Alt ←], [Alt →], [Alt ↓], [Alt ↑].
For huge sequences I prefer keyboard to mouse (which is "best practice" and "most relaxing")

Please note: Blur/Unblur procedure means browsing through dozens, hunderds or even thousands of images. Mapillary "work" is not "billable" is performed in my free time (which is both valuable and scarce).

> Next image having blurs set (filter on blurs)
> Seems rational at first but the intention and purpose of the blur editor is to not only correct false positives but also false negatives. A Jump to next image with blurs button would elude most users to only fix false positives, and perhaps only few false negatives which initially had some false positives in the same image.

My main issue:
After South Africa trip I had to upload 300.000 images. At first Webuploader was failing. After two weeks I decided to implement Python script. Another two days lead-time. By the time Python was more or less oké, Webuploader was also fixed. After one month Python also stopped uploading by incomplete Error Message "No handler could be found for logger exifread". So I had to use both Python Script and Webuploader.......
En than next step would be to check my images for blurs and unblurs?
Sorry, no way! This means being requested to build a house without any tools, or just using a hammer and saw...

Filter for images having blurs is not all bad idea. Although I see your point.
But also:
- Better to check for false positives - only - than to do nothing. Which is what I do now for 99% of my images.
Only 0,01% is used for mapping purposes in which as I do blur and unblur if I stumble into a blur.
Also, I sometimes (!) mark a GPS waypoint as a reminder: check for blurring faces! Which applies for 0,001% of my images.
- Filter on blurs will return images having blurs. These images are often related to other images close by having missing blurs. Often an image having one license plate or face blurred, will also have others missed. Often an image having blurs will be pointing to a complete sequence of images also having license plates and/or faces. So move to previous, previous, previous and then next, next, next, next... will check multiple images for blurs (triggered by one filtered image only).
It is unlikely (not impossible) that images not having any blurs are accompanied by others having missed blurs.

If Mapillary expects me to check my images for blurs/unblurs, I want tools.
Otherwise I won’t check...

Better suggestions as to resolve this issue would of course be appreciated.

**Comment by katrinhumal** [comment]
@AnkEric thanks for your very extensive and well thought through feedback. I definitely understand why you're requesting what you are. I'm not able to provide an equally detailed response right now but I've forwarded your feedback to the team and we'll do our best to take it into account. In broad strokes, the strategy is to make good efforts with the blurring algorithms so that the need for manual blurring would considerably decrease.

**Comment by ybkuang** [comment]
@AnkEric I hear your feedback and we have been keeping track closely of the issues that you mentioned. We are working the best we can to improve. 

- Python Upload Tools: We're tracing down the issues related to the exif libraries and setup dependency on different OS.  With that, we will make the tools easier for you to use. Sorry for the delay there. 

- Blurs:  We have been working heavily on that including data collection and training different version of the algorithms. The new algorithm addresses the issues (i.e. missing faces and false positives on useful map information) that you mentioned above. We are currently working on optimizing the throughput of algorithm before deploying it globally. We will see what we can do to help you before the global release is ready, to make your workflow easier. 

Again, thanks for your detailed feedback and suggestions.

**Comment by nereocystis** [comment]
I agree that a filter on blurs is a good idea, showing only images with blurs.

Also, it would be nice to have the total number of blurs in current image displayed on the side. Sometimes, I don't notice a blur.

My workflow for handling blurs is as follows:

1. Look through sequence for false positives, deleting this blurs.

2. As a separate workflow, I go through the sequence looking for missing blurs. This takes more time. As @AnkEric says, similar blurs tend to appear in sequence. In this workflow, I am looking for faces or license plates which should be blurred. Since the workflow is different, I prefer to handle the steps in a different pass.

**Comment by GITNE** [comment]
@AnkEric 
> That would be a solution. However [Alt ↑], [Alt ↓] does not act as expected in Google Chrome.

That's strange because it should work but I have not tested it. Please make sure you select the image into focus first by clicking on it, since (unfortunately) these keyboard shortcuts are not global anymore. Then you should be able to use the key shortcuts. @mapillary please check.

> Also, it would be nice to have the total number of blurs in current image displayed on the side. Sometimes, I don't notice a blur.

:+1: to that. And, I would also prefer to see this number in the stats than the changeset count, since a blur changeset with many blur fixes is sort of *more valuable* than a simple *clear all* changeset.

Generally speaking the blur editor could indeed use some usability improvements, as well as for pointing devices as for keyboards.

**Comment by AnkEric** [comment]
@GITNE: I'm still not capable of using [Alt ↑], [Alt ↓]. Not in Chrome, not in FireFox. 
LOL: I'm now (just now!) capable of using [Alt ↑], [Alt ↓].
Again:

> "Please make sure you select the image into focus first"

Now:

- Left hand is Keyboard:  [Alt ↑], [Alt ↓] (a little awkward).. 
- Right hand is Mouse: [Clear all +>] or [Submit +>]

But also: I finally discovered "the hare"...

![image](https://user-images.githubusercontent.com/20150550/35777811-a57e2492-09b4-11e8-906c-0e546fc17a70.png)
![image](https://user-images.githubusercontent.com/20150550/35777778-0370ae2c-09b4-11e8-99c3-13528bcbba7b.png)

Again, stupid, I never noticed the bar "|"...
For browsing, selecting, downloading my own images (for personal use) this is essential!

@Alessandro-Barbieri:

> I noticed that the algorithm blurs only perfectly readable and perfectly horizontal license plates.

You are so very correct... (I must admit to my regret).
And most parked cars don't show a  "perfectly horizontal license plate" (due to perspective).

@ybkuang:
In Durban (South Africa) I stumbled over a special case of "blurs".
Lots of cars, faces and buildings. Huge number of blurs.
Most typical: 10-20 false positives and only one licence plate, correctly blurred.
I was very tempting to Select [Clear all +>]. I didn't... (at least: i well behaved most of the time, except for mistakes).
But it would be very nice to have a "bounding rectangle" during selection of 10-20 false positives.

By example: select and delete all (false positives), except the "tent" (license plate or face)

![image](https://user-images.githubusercontent.com/20150550/35777983-7de576c6-09b7-11e8-8d1c-9901b40010f3.png)

**Comment by AnkEric** [comment]
Now I can reverse browsing order, nice. Sometime I do start at the end of a sequence...
But then, I also expect:
- [Clear all +>] to reverse into: [Clear all +<]
- [Submit +>] to reverse into: [Submit +<]

![image](https://user-images.githubusercontent.com/20150550/35779041-2e016c86-09c7-11e8-8508-dd7cd8b82517.png)

**Comment by AnkEric** [comment]
About Stats:
- It would be nice if I could find my own (un)blurs back again. I know I have cleared a license plate by mistake, but I cannot find it anymore...
- It would be nice if I could Bookmark an Image. Were did I stop processing last night?
- It would be nice if we all would be able to select sequence not yet (un)blurred. This might be considered as a ToDo list. And priority: don't blur/unblur a sequence that was already (more or less) processed.

**Comment by AnkEric** [comment]
I am impressed with Mapillary New Blur Algorithms.
I did not verify them all -)), but what I did see was impressive.
Even our kids walking ahead of us were blurred if a small (!) piece of face was visible from behind (!).
A group of twenty cyclists: all faces were blurred!
False positives: not found a single one of them.
Licences plates: all have been found, disregarding the angle and shape.
Compliments!
Thanks!
Case closed...!
